src/github.com/matrix-org/dendrite/clientapi/routing/register.go

```diff
@@ -425,18 +425,21 @@ func checkFlows(a []authtypes.LoginType, b []authtypes.LoginType) bool {
 	sort.Slice(a, func(i, j int) bool { return a[i] < a[j] })
```

Comment:
hrm. This is going to modify the inputs to `checkFlows` - ie, it's going to write to `cfg.Derived.Registration.Flows` and `sessions[sessionID]`.

I guess it's harmless enough, but it's certainly unexpected. It also highlights the redundancy here in sorting for every pairing of session/flow.

My suggestion would be that you do one of:
* sort the lists much earlier and require sorted lists be passed into `checkFlowCompleted`
* rather than sorting the lists, construct a `map[LoginType]bool` for the completed stages and iterate over the steps in `b`

While I'm here, can we come up with better names than `a` and `b`? `completedStages` / `requiredStages` ?

Reply:
Would simply creating a copy of the stages before sorting them in-place be sufficient?